docs/design/tracking-branches.md

```diff
@@ -0,0 +1,299 @@
+# Remote/`@git` tracking branches
+
+This is a plan to implement more Git-like remote tracking branch UX.
+
+## Objective
+
+`jj` imports all remote branches to local branches by default. As described in
+[#1136], this doesn't interact nicely with Git if we have multiple Git remotes
+with a number of branches. The `git.auto-local-branch` config can mitigate this
+problem, but we'll get locally-deleted branches instead.
+
+The goal of this plan is to implement
+* proper support for tracking/non-tracking remote branches
+* logically consistent data model for importing/exporting Git refs
+
+[#1136]: https://github.com/martinvonz/jj/issues/1136
+
+## Current data model (as of jj 0.8.0)
+
+Under the current model, all remote branches are "tracking" branches, and
+remote changes are merged into the local counterparts.
+
+```
+branches
+  [name]:
+    local_target?
+    remote_targets[remote]: target
+tags
+  [name]: target
+git_refs
+  ["refs/heads/{name}"]: target             # last-known local branches
+  ["refs/remotes/{remote}/{name}"]: target  # last-known remote branches
+                                            # (copied to remote_targets)
+  ["refs/tags/{name}"]: target              # last-known tags
+git_head: target?
+```
+
+* Remote branches are stored in both `branches[name].remote_targets` and
+  `git_refs["refs/remotes"]`. These two are kept in sync unless the branch is
+  removed by `jj branch forget` command.
+* Pseudo `@git` remote branches are stored in `git_refs["refs/heads"]`.
+
+## Proposed data model
+
+We'll add a per-remote-branch `state` to distinguish non-tracking branches
+from tracking ones.
+
+```
+state = new        # not merged in the local branch or tag
+      | tracking   # merged in the local branch or tag
+      | forgotten  # to be expunged on the next export
+# `ignored` state could be added if we want to manage it by view, not by
+# config file. target of ignored remote branch would be absent.
+```
```

Comment:
I haven't thought very deeply about how/whether this fits into the design, but as a user I might want even more states (not necessarily in the first version):

- I might want a "private" state, where a branch is not going to be pushed to this remote. So, I could push "private" branches to my Github fork but not to the main repo.
- We could support git-like differently-named branches on the remote by storing the name of the remote branch (for this remote) that this local branch corresponds to in this "state" or next to it.

Reply:
>     * I might want a "private" state, where a branch is not going to be pushed to this remote. So, I could push "private" branches to my Github fork but not to the main repo.

I haven't considered this, but it might be a state to be set to a local branch, meaning don't push this branch to untraked remote. It also resembles Mercurial's secret phase, but a bit different since secret heads can't be pushed to any remotes.

>     * We could support git-like differently-named branches on the remote (as [recently requested on Discord](https://discord.com/channels/968932220549103686/969829516539228222/1140657566511288330)) by storing the name of the remote branch (for this remote) that this local branch corresponds to in this "state" or next to it.

Yeah, it could be something like `State::Tracking { local_branch_name: String }`. I'm not sure how much that will complicate the UI.

Reply:
Marked as resolved. I've added these to "remaining issues"